src/scheduler/check.cpp

```diff
@@ -1570,28 +1570,43 @@ check_normal_node_path(status *policy, server_info *sinfo, queue_info *qinfo, re
 			/* if there are nodes assigned to the queue, then check those */
 			if (qinfo->has_nodes)
 				ninfo_arr = qinfo->nodes;
-			else if (!(sinfo->svr_to_psets.empty())) {	/* Multi-server psets */
-				/* Find the pset of the server which owns the job */
-				for (auto &spset: sinfo->svr_to_psets) {
-					if (spset.svr_inst_id == resresv->job->svr_inst_id) {
-						msvr_pset[0] = spset.np;
-						if (!resresv->job->is_array && spec->total_chunks == 1) {
-							/* Allow job to be placed on nodes of other servers as well
-							 * For now, restricting job arrays and multi-chunk jobs
-							 * to nodes of just the local server
-							 */
-							msvr_pset[1] = sinfo->allpart;
-							msvr_pset[2] = NULL;
-						} else {
-							/* This prevents eval_selspec from considering all unassociated nodes */
-							flags &= ~SPAN_PSETS;
-							msvr_pset[1] = NULL;
-						}
-						nodepart = msvr_pset;
-						break;
+		}
+
+		if (!(sinfo->svr_to_psets.empty())) { /* Multi-server psets */
+			int i = 0;
+			int owner_set = -1;
+
+			/* Find the pset of the server which owns the job */
+			for (auto &spset : sinfo->svr_to_psets) {
+				if (spset.svr_inst_id == resresv->svr_inst_id) {
+					msvr_pset[0] = spset.np;
+					owner_set = i;
+					break;
+				}
+				i++;
+			}
+			if (owner_set != -1) {
+				/* Restrict job arrays and reservations to owner server */
+				if (resresv->job->is_array || resresv->is_resv)
+					msvr_pset[1] = NULL;
+				else if (spec->total_chunks == 1) {    /* Single chunk jobs */
```

Comment:
In case of single chunk jobs if owner nodes do not work, cannot we try for other partitions like the way we do for multi-node jobs ?

Reply:
we don't need to go partition by partition for single chunk jobs as they won't span. So, we can directly try ALL nodes if the  owner server's nodes don't work.